include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
 	const float ay = params_getAlphaV(params);
 	const float ay2 = ay*ay;
 	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
-	const bool is_bsdf = !op_isBRDF(op);
-	const vec3 refl = params_getReflectance(params);
-
-#ifndef NO_TWOSIDED
-	bool ts_flag = false;
-	handleTwosided_interactionOnly(ts_flag, instr);
-#endif //NO_TWOSIDED
+	const mat2x3 ior2 = matrixCompMult(ior,ior);
+	const bool is_bsdf = !op_isBRDF(op) && !is_coat;
+	const vec3 albedo = params_getReflectance(params);
 
+	float pdf = 1.0;
+	vec3 rem = vec3(1.0);
 	uint ndf = instr_getNDF(instr);
-	irr_glsl_BSDFSample s;
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		s = irr_glsl_oren_nayar_cos_generate(currInteraction, rand.xy, ax2);
-	} else 
-#endif //OP_DIFFUSE
+	irr_glsl_LightSample s;
 
-#ifdef OP_DIFFTRANS
-	if (op==OP_DIFFTRANS) {
-		//TODO take into account full sphere
-		s = irr_glsl_cos_weighted_cos_generate(currInteraction, rand.xy);
+	const vec3 localV = irr_glsl_getTangentSpaceV(currInteraction);
+	const mat3 tangentFrame = irr_glsl_getTangentFrame(currInteraction);
+
+#ifdef OP_DELTATRANS
+	if (op == OP_DELTATRANS)
+	{
+		s = irr_glsl_createLightSample(-precomp.V, -1.0, currInteraction.T, currInteraction.B, currInteraction.isotropic.N);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		rem = vec3(1.0);
+		pdf = irr_glsl_FLT_INF;
 	} else
-#endif //OP_DIFFTRANS
+#endif
+#ifdef OP_THINDIELECTRIC
+	if (op == OP_THINDIELECTRIC)
+	{
+		const vec3 luminosityContributionHint = CIE_XYZ_Luma_Y_coeffs;
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior2[0], luminosityContributionHint);
```

Comment:
the most complicated form of the function actually cuts down on computation
https://github.com/Devsh-Graphics-Programming/Nabla/blob/881deedeac39b6141d061756fa9b00951fb57f28/include/irr/builtin/glsl/bxdf/bsdf/specular/dielectric.glsl#L17

returns remainder metadata, which you can put into this
https://github.com/Devsh-Graphics-Programming/Nabla/blob/881deedeac39b6141d061756fa9b00951fb57f28/include/irr/builtin/glsl/bxdf/bsdf/specular/dielectric.glsl#L48